rc/filetype/rust.kak

```diff
@@ -34,9 +34,34 @@ add-highlighter shared/rust regions
 add-highlighter shared/rust/code default-region group
 add-highlighter shared/rust/string           region %{(?<!')"} (?<!\\)(\\\\)*"              fill string
 add-highlighter shared/rust/raw_string       region -match-capture %{(?<!')r(#*)"} %{"(#*)} fill string
-add-highlighter shared/rust/comment          region -recurse "/\*" "/\*" "\*/"              fill comment
-add-highlighter shared/rust/documentation    region "//[!/]" "$"                            fill documentation
-add-highlighter shared/rust/line_comment     region "//" "$"                                fill comment
+
+add-highlighter shared/rust/line_doctest region ^\h*//[!/]\h*```($|should_panic|no_run|ignore|allow_fail|rust|test_harness|compile_fail|E\d{4}|edition201[58]) ^\h*//[!/]\h*```$ regions
+add-highlighter shared/rust/line_doctest/marker region ```.* $ group
+add-highlighter shared/rust/line_doctest/marker/fence regex ``` 0:meta
+add-highlighter shared/rust/line_doctest/marker/keywords regex [\d\w] 0:meta # already matched above, just ignore comma
+add-highlighter shared/rust/line_doctest/inner region '^\h*//[!/]\h+#' ' ' group
+add-highlighter shared/rust/line_doctest/inner/comment regex //[!/] 0:documentation
+add-highlighter shared/rust/line_doctest/inner/hidden regex '#' 0:meta
+add-highlighter shared/rust/line_doctest/comment region ^\h*//[!/] ' '                      fill documentation
+add-highlighter shared/rust/line_doctest/code default-region ref rust
+add-highlighter shared/rust/line_code_rest   region ^\h*//[!/]\h*``` ^\h*//[!/]\h*```$      fill documentation # reset invalid doctest
+add-highlighter shared/rust/line_comment2    region //[!/]{2} $                             fill comment
+add-highlighter shared/rust/line_doc         region //[!/] $                                fill documentation
+add-highlighter shared/rust/line_comment1    region // $                                    fill comment
+
+add-highlighter shared/rust/block_comment2   region -recurse /\*\*\* /\*\*\* \*/            fill comment
+add-highlighter shared/rust/block_doc        region -recurse /\*\* /\*\* \*/ regions
+add-highlighter shared/rust/block_doc/doctest region ```($|should_panic|no_run|ignore|allow_fail|rust|test_harness|compile_fail|E\d{4}|edition201[58]) ```$ regions
+add-highlighter shared/rust/block_doc/doctest/marker region ```.* $ group
+add-highlighter shared/rust/block_doc/doctest/marker/fence regex ``` 0:meta
+add-highlighter shared/rust/block_doc/doctest/marker/keywords regex [\d\w] 0:meta # already matched above, just ignore comma
+add-highlighter shared/rust/block_doc/doctest/inner default-region group
+add-highlighter shared/rust/block_doc/doctest/inner/hidden regex '^\h*\**\h*#' 0:meta
+add-highlighter shared/rust/block_doc/doctest/inner/comment regex ^\h*\* 0:documentation
+add-highlighter shared/rust/block_doc/doctest/inner/code ref rust
+add-highlighter shared/rust/block_doc/code_rest region ``` ``` fill documentation
```

Comment:
Looks like `code_rest` and `doctest` are competing regions, one highlighting the background of the other.

Could you try unifying those?

Reply:
No, I do want them to compete, `code_rest` is supposed to be those code blocks that are not matched, if not it will have weird behavior.
```rust
/// ```python
/// print("Hello")
/// ``` <- this will not end
/// <- this will be highlighted as rust syntax
/// ```rust <- continues to be highlighted
/// <- continues to be highlighted
/// ``` <- it only ends here
```